File: Tests/TrendModelTests.swift
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

import XCTest
@testable import Doigt

/// History's trend deck, now built off the main actor from copied rows. These pin the
/// rules it carried over from the view: which routines get a card and under what name,
/// which grips a card offers and in what order, and what one point on a line means.
final class TrendModelTests: XCTestCase {
    private let crimp = GripSpec(edgeMM: 20, fingers: .four, position: .halfCrimp)
    private let front3 = GripSpec(edgeMM: 20, fingers: FingerSet(token: "IMR"), position: .halfCrimp)
    private let base = Date(timeIntervalSinceReferenceDate: 800_000_000)

    private func rep(_ grip: GripSpec, kg: Double, held: Double = 10,
                     outcome: RepOutcome = .completed) -> RepSummary {
        var rep = RepSummary()
        rep.grip = grip
        rep.avgKg = kg
        rep.peakKg = kg + 2
        rep.heldSeconds = held
        rep.outcome = outcome
        return rep
    }

    private func row(_ reps: [RepSummary], routine: UUID?, name: String = "Frozen",
                     daysAgo: Double) -> TrendModel.Row {
        TrendModel.Row(id: UUID(), templateID: routine, templateName: name,
                       startedAt: base.addingTimeInterval(-daysAgo * 86_400),
                       resultsData: BlobCodec.encode(reps) ?? Data())
    }

    func testCardsFollowTheMostRecentlyTrainedRoutineAndItsLiveName() {
        let daily = UUID(), maxDay = UUID()
        let rows = [   // newest first, as History's query sorts them
            row([rep(crimp, kg: 30)], routine: maxDay, name: "Old max name", daysAgo: 1),
            row([rep(crimp, kg: 12)], routine: daily, daysAgo: 2),
            row([rep(crimp, kg: 11)], routine: daily, daysAgo: 3),
        ]
        let model = TrendModel.build(rows: rows, routineNames: [daily: "Daily", maxDay: "Max day"])
        XCTAssertEqual(model.routines.map(\.name), ["Max day", "Daily"])
        XCTAssertEqual(model.routines.map(\.key), [maxDay.uuidString, daily.uuidString])
    }

    func testADeletedRoutineGetsNoCardAndANamelessOneKeepsItsFrozenName() {
        let gone = UUID()
        let rows = [
            row([rep(crimp, kg: 12)], routine: gone, daysAgo: 1),
            row([rep(crimp, kg: 12)], routine: nil, name: "Before IDs", daysAgo: 2),
        ]
        let model = TrendModel.build(rows: rows, routineNames: [:])
        XCTAssertEqual(model.routines.map(\.name), ["Before IDs"])
    }

    /// Gauge-free sessions log 0 kg and skipped reps never finished; neither is a load.
    func testOnlyMeasuredCompletedRepsChart() {
        let daily = UUID()
        let rows = [
            row([rep(crimp, kg: 0)], routine: daily, daysAgo: 1),
            row([rep(crimp, kg: 14, outcome: .skipped)], routine: daily, daysAgo: 2),
        ]
        XCTAssertTrue(TrendModel.build(rows: rows, routineNames: [daily: "Daily"]).routines.isEmpty,
                      "a routine with nothing measured gets no card")
    }

    func testGripsAreMostTrainedFirstAndSeriesAreTimeWeightedOldestFirst() throws {
        let daily = UUID()
        let rows = [
            row([rep(front3, kg: 9), rep(crimp, kg: 20, held: 10), rep(crimp, kg: 10, held: 30)],
                routine: daily, daysAgo: 1),
            row([rep(crimp, kg: 12)], routine: daily, daysAgo: 5),
        ]
        let routine = try XCTUnwrap(TrendModel.build(rows: rows, routineNames: [daily: "Daily"])
            .routines.first)
        XCTAssertEqual(routine.grips.map(\.key), [crimp.key, front3.key])
        XCTAssertEqual(routine.grips.map(\.count), [3, 1])
        let series = try XCTUnwrap(routine.series[crimp.key])
        XCTAssertEqual(series.map(\.avgKg), [12, 12.5],
                       "oldest first; (20·10 + 10·30) / 40 for the newer session")
        XCTAssertEqual(series.map(\.id), [rows[1].id, rows[0].id])
    }

    func testASharedSelectionHoldsOnlyWhereTheRoutineTrainedIt() throws {
        let daily = UUID()
        let routine = try XCTUnwrap(TrendModel.build(
            rows: [row([rep(crimp, kg: 12)], routine: daily, daysAgo: 1)],
            routineNames: [daily: "Daily"]).routines.first)
        XCTAssertEqual(routine.selectedGrip(nil), crimp.key)
        XCTAssertEqual(routine.selectedGrip(front3.key), crimp.key)
        XCTAssertEqual(routine.selectedGrip(crimp.key), crimp.key)
    }
}

File: android/app/src/test/kotlin/run/nuri/getagrip/ui/TrendModelTests.kt
// SPDX-License-Identifier: MPL-2.0
// Original contributions Copyright 2026 Nuri Bruner.

package run.nuri.getagrip.ui

import run.nuri.getagrip.data.WorkoutLogEntity
import run.nuri.getagrip.engine.BlobCodec
import run.nuri.getagrip.engine.FingerSet
import run.nuri.getagrip.engine.GripPosition
import run.nuri.getagrip.engine.GripSpec
import run.nuri.getagrip.engine.RepOutcome
import run.nuri.getagrip.engine.RepSummary
import run.nuri.getagrip.engine.SessionKind
import run.nuri.getagrip.ui.history.TrendModel
import java.time.Instant
import java.util.UUID
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertIs
import kotlin.test.assertNull
import kotlin.test.assertTrue

/// History's trend deck, built off the main thread from the feed's logs. The twin of iOS's
/// `TrendModelTests`: which routines get a card and under what name, which grips a card
/// offers and in what order, and what one point on a line means.
class TrendModelTests {
    private val crimp = GripSpec(20, FingerSet.four, GripPosition.halfCrimp)
    private val front3 = GripSpec(20, FingerSet.fromToken("IMR"), GripPosition.halfCrimp)
    private val base = Instant.parse("2026-05-10T08:00:00Z")

    private fun rep(grip: GripSpec, kg: Double, held: Double = 10.0, outcome: RepOutcome = RepOutcome.completed) =
        RepSummary(grip = grip, avgKg = kg, peakKg = kg + 2, heldSeconds = held, outcome = outcome)

    private fun row(
        reps: List<RepSummary>,
        routine: UUID?,
        name: String = "Frozen",
        daysAgo: Double,
        kind: SessionKind = SessionKind.hang,
    ) = WorkoutLogEntity(
        templateID = routine,
        templateName = name,
        startedAt = base.minusSeconds((daysAgo * 86_400).toLong()),
        resultsData = BlobCodec.encodeAll(reps) ?: "",
        kindRaw = kind.rawValue,
    )

    @Test fun cardsFollowTheMostRecentlyTrainedRoutineAndItsLiveName() {
        val daily = UUID.randomUUID()
        val maxDay = UUID.randomUUID()
        val rows = listOf( // newest first, as the feed sorts them
            row(listOf(rep(crimp, 30.0)), maxDay, name = "Old max name", daysAgo = 1.0),
            row(listOf(rep(crimp, 12.0)), daily, daysAgo = 2.0),
            row(listOf(rep(crimp, 11.0)), daily, daysAgo = 3.0),
        )
        val model = TrendModel.build(rows, mapOf(daily to "Daily", maxDay to "Max day"))
        assertEquals(listOf("Max day", "Daily"), model.routines.map { it.name })
        assertEquals(
            listOf(maxDay.toString().uppercase(), daily.toString().uppercase()),
            model.routines.map { it.key },
        )
    }

    @Test fun aDeletedRoutineGetsNoCardAndANamelessOneKeepsItsFrozenName() {
        val gone = UUID.randomUUID()
        val rows = listOf(
            row(listOf(rep(crimp, 12.0)), gone, daysAgo = 1.0),
            row(listOf(rep(crimp, 12.0)), null, name = "Before IDs", daysAgo = 2.0),
        )
        assertEquals(listOf("Before IDs"), TrendModel.build(rows, emptyMap()).routines.map { it.name })
    }

    /// Gauge-free sessions log 0 kg and skipped reps never finished; neither is a load.
    @Test fun onlyMeasuredCompletedRepsChart() {
        val daily = UUID.randomUUID()
        val rows = listOf(
            row(listOf(rep(crimp, 0.0)), daily, daysAgo = 1.0),
            row(listOf(rep(crimp, 14.0, outcome = RepOutcome.skipped)), daily, daysAgo = 2.0),
        )
        assertTrue(
            TrendModel.build(rows, mapOf(daily to "Daily")).routines.isEmpty(),
            "a routine with nothing measured gets no card",
        )
    }

    /// A climb or a hand-logged hang carries no reps anyway; the kind filter says so outright.
    @Test fun onlyHangSessionsAreRead() {
        val rows = listOf(row(listOf(rep(crimp, 12.0)), null, name = "Gym", daysAgo = 1.0, kind = SessionKind.climbLimit))
        assertTrue(TrendModel.build(rows, emptyMap()).routines.isEmpty())
    }

    @Test fun gripsAreMostTrainedFirstAndSeriesAreTimeWeightedOldestFirst() {
        val daily = UUID.randomUUID()
        val rows = listOf(
            row(listOf(rep(front3, 9.0), rep(crimp, 20.0, held = 10.0), rep(crimp, 10.0, held = 30.0)), daily, daysAgo = 1.0),
            row(listOf(rep(crimp, 12.0)), daily, daysAgo = 5.0),
        )
        val routine = TrendModel.build(rows, mapOf(daily to "Daily")).routines.single()
        assertEquals(listOf(crimp.key, front3.key), routine.grips.map { it.key })
        assertEquals(listOf(3, 1), routine.grips.map { it.count })
        val series = routine.series.getValue(crimp.key)
        // Oldest first; (20·10 + 10·30) / 40 for the newer session.
        assertEquals(listOf(12.0, 12.5), series.map { it.avgKg })
        assertEquals(listOf(rows[1].id, rows[0].id), series.map { it.id })
    }

    @Test fun aSharedSelectionHoldsOnlyWhereTheRoutineTrainedIt() {
        val daily = UUID.randomUUID()
        val routine = TrendModel.build(
            listOf(row(listOf(rep(crimp, 12.0)), daily, daysAgo = 1.0)),
            mapOf(daily to "Daily"),
        ).routines.single()
        assertEquals(crimp.key, routine.selectedGrip(null))
        assertEquals(crimp.key, routine.selectedGrip(front3.key))
        assertEquals(crimp.key, routine.selectedGrip(crimp.key))
    }

    /// The feed's cache is what the deck reads through, so the build must never go around it.
    @Test fun repsComeThroughTheSuppliedReader() {
        val daily = UUID.randomUUID()
        val log = row(emptyList(), daily, daysAgo = 1.0)
        var asked = 0
        val model = TrendModel.build(listOf(log), mapOf(daily to "Daily"), reps = {
            asked++
            listOf(rep(crimp, 15.0))
        })
        assertEquals(1, asked)
        assertEquals(15.0, model.routines.single().series.getValue(crimp.key).single().avgKg)
    }

    @Test fun aCancelledBuildPublishesNothing() {
        val daily = UUID.randomUUID()
        val rows = listOf(row(listOf(rep(crimp, 12.0)), daily, daysAgo = 1.0))
        assertTrue(TrendModel.build(rows, mapOf(daily to "Daily"), isCancelled = { true }).routines.isEmpty())
    }

    /// Under half a kilo across a series is grip noise; calling it progress would be flattery.
    @Test fun theSummaryCallsHalfAKiloNoise() {
        fun series(vararg kg: Double) = kg.mapIndexed { i, v -> TrendModel.Point(UUID.randomUUID(), base.plusSeconds(i * 86_400L), v) }
        assertNull(TrendModel.summary(emptyList()))
        assertEquals(TrendModel.Summary.Steady(12.4), TrendModel.summary(series(12.0, 13.0, 12.4)))
        val up = assertIs<TrendModel.Summary.Up>(TrendModel.summary(series(12.0, 12.2, 13.0)))
        assertEquals(1.0, up.deltaKg, 1e-9)
        assertEquals(3, up.sessions)
        val down = assertIs<TrendModel.Summary.Down>(TrendModel.summary(series(14.0, 13.5)))
        assertEquals(0.5, down.deltaKg, 1e-9)
        assertEquals(2, down.sessions)
    }
}
